WIP: Starting to update the PEP to include SCC + Atomic RC + Big Picture (#10)
TobiasWrigstad/peps: peps/pep-0795.rst

========

This PEP proposes adding a mechanism for deep immutability to
Python. Our main motivation for wanting to add immutability to
Python is to facilitate concurrent programming, both using
free-threaded Python and subinterpreters. The immutability this
PEP proposes will make it possible for subinterpreters to share
immutable objects by reference. In free-threaded Python, it will
be possible to avoid some object locking overhead when interacting
with immutable objects, although we are not proposing to do the
optimisations for free-threaded Python as part of this PEP.

> The mechanism requires some changes to the core language, but
> user-facing functions are delivered in a module called
> ``immutable``. This module provides the following functions and
> types:

Reviewer: This sounds like we just packaged a core feature into a library, how about:

```suggestion
Immutabiliy requires some changes to the code language, but
these changes should only effect programs, if a user engages with
the feature. The user-facing functions are therefore delivered
as a module called ``immutable``. This module provides the
following functions and types:
```

This PEP proposes adding a mechanism for deep immutability to

> Python. Our main motivation for wanting to add immutability to

Reviewer: Assuming that we expand the scope of the pep, this intro should also reflect this: "and a way to share immutable data by reference across subinterpreters"

for subinterpreters to share immutable objects directly by
reference, which is a big performance benefit.

This change requires modules loaded in one subinterpreter to be

> accessible from another.

Reviewer: This might no longer be true, if we restrict the types which can be frozen.

If this sentence it kept, there should be an explanation why as this might not be obvious for people just getting into this topic.

*shallow* immutable objects (``tuple``, ``frozenset``).

The Big Picture
===============

Note: this text goes far beyond what this PEP intends to deliver,
in order to clarify and motivate some of our design decisions.
There are lots of details left out below, including a discussion
about escape hatches, etc. These will be addressed in the next
PEP.

This PEP builds towards a data-race free programming model for
Python. What we envision beyond this PEP is a Python that does not
permit concurrent mutation, meaning it should not be possible for
two threads to mutate an object at the same time. The same
property falls out of properly synchronised programs, but poorly
synchronised programs may contain data races that silently corrupt
data, or crashes the the interpreter. We want such programs to
raise exceptions instead to avoid these errors and also point
programmers to how to fix them. Ensuring that a program is
properly synchronised is hard, and debugging things like data

> races is often very challenging.

Reviewer: NIT: This last sentence feels like it was just tagged on. Wasn't this to a weaker extent said earlier in this paragraph?

ability to create groups of mutable objects which we call
*regions*. Regions are *isolated*, which we for now will define as
"objects inside the region only point to other objects inside the
same region". We will relax this shortly, once we get around to
introduce more of the moving pieces. For now, all objects in a
Python program belong to some region.

A region is owned by a thread and unless a thread owns a region,
it cannot access the objects inside it. If you want to share a
region between threads it must be wrapped inside a lock, which
ensures that only one thread at a time owns the region. Trying to
release a lock for a region while hanging on to direct references
into the region throws an exception. (Otherwise, two threads could
end up mutating the same object at the same time.) In the rest of
this section, ”lock” refers to a lock object that manages the
ownership of a single region. (It is possible to extend this model
with a read-write lock which permits multiple threads to read
objects at the same time, but not mutate them.)

The immutable objects that this PEP proposes are safe to share
across different threads (or subinterpreters) without leading to

> concurrent mutation. As the purpose of regions is to coordinate
> access to mutable objects, immutable objects cannot live in
> regions. Furthermore, objects in regions are permitted to
> reference immutable objects. In the regions model, we believe that

Reviewer: How about?

```suggestion
concurrent mutation. The purpose of regions is to coordinate
access to mutable objects. Immutable objects don't require regions
to be savely shared, since they're immutable. Furthermore, our model
permits references from regions to immutable objects. In the regions
model, we believe that
```

immutable objects are a great way for regions to share common

> data. One example of common data is *types*.

Reviewer: This feels like an "are" but could be an "is" since it is singular "data"

```suggestion
data. One example of common data are *types*.
```

If regions are isolated, and isolation means that all pointers in
an object point to other objects in the same region, then since an
object has a pointer to its type, the type must reside in the same
region. That's quite restrictive, and probably not what you want
in a program that makes heavy use of regions for concurrency. One

> way around this is to make the type immutable.

Reviewer: This is a bit confusing on a quick read, maybe make it more explicit. Something like:

"Each object has a pointer to its type. When an object is moved into a region, the type will also be moved. However, this would restrict all objects of this type from being moved into another region, as that would break isolation"

To simplify this process, we propose that certain types of objects
should be possible to make immutable implicitly. For example, as
soon as a type object becomes shared between two regions, we make
it immutable. Objects that seem sensible to freeze implicitly

> include types, strings, tuples, and numbers.

Reviewer: Do you mean tuples in general or specifically?
```suggestion
include types, strings, numbers, and tuples of implicitly freezable types.
```

Regions with locks add an escape mechanism to immutability that
this PEP does not discuss. Freezing does not propagate through
locks. Thus, if a programmer wants an immutable object to contain
some mutable state — for example, a counter inside an immutable

> class that counts how many instances it has, they can achieve this

Reviewer: This example started with a dash
```suggestion
class that counts how many instances it has - they can achieve this
```

by sticking a lock with a region inside in the immutable class to
hold the counter.

When an object is created, it lives in the "default region" of the
current thread. This is essentially a thread-local storage which
also contains all the stack frames of that thread. Variables and
objects in the default region are permitted to point into the

> regions that the thread owns. This means that objects are
> thread-local when created.

Reviewer: I see how the second thing is an implication, but maybe make it as a statement

```suggestion
regions that the thread owns. All objects are
thread-local when they are created.
```

opt-in model is that a large part of all Python libraries cannot
be (even nominally) made immutable (out-of-the-box).

Because immutability is deep, and concerns have been raised about

> the propagation of deep immutability, we propose an opt-in model.

Reviewer: Maybe also mention:

We considered shallow immutability as PEP X suggested, but determined that this is not sufficient. The reasons are discussed later in this pep

Consequences of Deep Freezing
=============================

* The most obvious consequence of deep freezing is that it can
  lead to surprising results when programmers fail to reason
  correctly about the object structures in memory and how the
  objects reference each other. For example, consider
  ``freeze(x)`` followed by ``y.f = 42``. If the object in ``x``
  can reach the same object that ``y`` points to, then, the
  assignment will fail. **Mitigation:** Support for freezing is
  opt-in. Further, to facilitate debugging, exceptions due to
  attempting to mutate immutable objects will include information

>   about on what line an object was made immutable.

Reviewer: Currently, this is a debugging option, and we might want to keep this as only a debugging flag to avoid the overhead this adds. So maybe make this conditional on the debugging flag
